### DISCUSSION TITLE
Advice on creating model

### DISCUSSION THREAD

**Discussion by MagnusMyhre** [IDAES/idaes-pse#764] [Q&A] [answered]
Hi

I am to model a liquefaction of natural gas process like shown below.
![Screenshot (10)](https://user-images.githubusercontent.com/77294919/159538854-86f452d7-5d19-4613-a7cc-75a489faedc8.png)
And here are the inputs:
<img width="445" alt="Input (2)" src="https://user-images.githubusercontent.com/77294919/159543779-e231cd45-ea74-4d9b-8dea-ef7e6aa03105.png">

The model will contain two streams, one refrigeration loop and one natural gas/liquid stream. 
The refrigeration stream will be of a gas/liquid state containing all the components as in the natural gas/liquid stream except for one component. 

I was wondering if you could help me with how I should approach this. What existing property package could be smart to perhaps configurate and what properties and methods should I use to account for e.g phase equillibrium and so on. Also what heat exchanger should I pick and how should I model it as it may be a bit tricky since it contains a loop?

Thank you

Magnus

**Comment by andrewlee94** [comment, marked as answer]
@agarciadiego Might be able to offer some insight on this. I believe he has developed a property package for mixed hydrocarbons using the Pen-Robinson EoS that might be exactly what you need for this.

The one thing that I can see that IDAES is missing right now is the three stream heat exchanger, for which you wound need to develop your own model. However, we we would be interested in seeing the model you come up with hand possibly adding it to the IDAES model libraries.

**Reply by agarciadiego** [reply]
@MagnusMyhre We have an example of the generic property package for light hydrocarbons (https://github.com/IDAES/idaes-pse/blob/main/idaes/generic_models/properties/core/examples/HC_PR.py), and there is one that utilizes nitrogen (https://github.com/IDAES/idaes-pse/blob/main/idaes/generic_models/properties/core/examples/ASU_PR.py). In examples-pse we have a couple of jupyter notebooks where these are used (https://github.com/IDAES/examples-pse/tree/3155f1efe4eb3635dfbbd5395ccc0aeed0c41913/src/_skip)

**Comment by MagnusMyhre** [comment]
Thank you guys, much appreciated. I will try to look at how to create the heat exchanger.

Kind regards
Magnus

**Reply by andrewlee94** [reply]
Whilst looking through some files today I cam across this, which might be useful: https://github.com/IDAES/idaes-pse/blob/main/idaes/power_generation/unit_models/heat_exchanger_3streams.py

Note however that this is about to be moved into `idaes.models_extra.power_generation`.

**Reply by MagnusMyhre** [reply]
That will come in handy, thanks a lot!

**Comment by MagnusMyhre** [comment]
Hi again. 

I realized that it was a little bit complex and time consuming to develop my own LNG heat exchanger. However I found another method that use two shell and tube heat exchangers in a way that mimics a multistream heat exchanger. I refer to the following article:
[MSHE.pdf](https://github.com/IDAES/idaes-pse/files/8542234/MSHE.pdf)

I took your advice on creating my property package from the already existing hydrocarbon package. I've changed and added all the necessary properties and values for the task at hand (I think at least). 

I created my IDAES model and connected all my unit models and transformed the arcs with TransformationFactory. 
My problems start now with the specification and initialization of the model.
I added my conditions to the model, and the degrees of freedom are unfortunately -5. I used the sequential decomposition tool and added the tear guesses where I saw fit. 

The initialization started and found optimal results all the way until the heat exchanger. It seems like there are some scaling problems from the output logs. 

Anyway, attached is the Property package, Jupyter notebook IDAES model, Output log (DEBUG) from my last attempt at initialization, and pictures showing the model in HYSYS and by the visualization tool in IDAES and finally the article referred to above. 

[IDAES_model.zip](https://github.com/IDAES/idaes-pse/files/8542368/IDAES_model.zip)

Could you perhaps take a look and if possible come up with a solution to how I can get my model to initialize and also how I achieve a square problem so there are no problems with the degrees of freedom. I appreciate any help to move forward on this. 

Kind regards, 
Magnus

**Comment by andrewlee94** [comment]
Hi Magnus,

I took a quick look at your flowsheet, and whilst I wasn’t able to work out all of what is going on, however the first thing that is obvious is that you initially had 15 degrees of freedom but that you fixed 20 operating conditions. Thus, it is not surprising that you have -5 DoF after this.

Looking at the diagram of the system, I think I see the problem: the tube side seems to be completely cyclic (i.e. there are inlets/outlets to this cycle). This means that the entire loop is self-referential – by setting the conditions at any point in the cycle this defines the conditions at every other point in the cycle and eventually cycles back around to implicitly defining that same variables you just fixed. Hence, you cannot define any conditions in that cycle without over-specifying the system. What you need to do is break the cycle at some point (i.e. leave out one of the Arcs), and then treat the break point as an inlet. Given the completely cyclic nature of your system, the “outlet” at the other end of the break should end up being identical to the conditions at the inlet, but you should verify this to be sure. To me, the most obvious point to put this break would be between the Mixer and Compressor unit.

Andrew


From: MagnusMyhre ***@***.***>
Sent: Friday, April 22, 2022 12:04 PM
To: IDAES/idaes-pse ***@***.***>
Cc: Lee, Andrew (FN) ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [IDAES/idaes-pse] Advice on creating model (Discussion #764)


Hi again.

I realized that it was a little bit complex and time consuming to develop my own LNG heat exchanger. However I found another method that use two shell and tube heat exchangers in a way that mimics a multistream heat exchanger. I refer to the following article:
MSHE.pdf<https://github.com/IDAES/idaes-pse/files/8542234/MSHE.pdf>

I took your advice on creating my property package from the already existing hydrocarbon package. I've changed and added all the necessary properties and values for the task at hand (I think at least).

I created my IDAES model and connected all my unit models and transformed the arcs with TransformationFactory.
My problems start now with the specification and initialization of the model.
I added my conditions to the model, and the degrees of freedom are unfortunately -5. I used the sequential decomposition tool and added the tear guesses where I saw fit.

The initialization started and found optimal results all the way until the heat exchanger. It seems like there are some scaling problems from the output logs.

Anyway, attached is the Property package, Jupyter notebook IDAES model, Output log (DEBUG) from my last attempt at initialization, and pictures showing the model in HYSYS and by the visualization tool in IDAES and finally the article referred to above.

IDAES_model.zip<https://github.com/IDAES/idaes-pse/files/8542368/IDAES_model.zip>

Could you perhaps take a look and if possible come up with a solution to how I can get my model to initialize and also how I achieve a square problem so there are no problems with the degrees of freedom. I appreciate any help to move forward on this.

Kind regards,
Magnus

—
Reply to this email directly, view it on GitHub<https://github.com/IDAES/idaes-pse/discussions/764#discussioncomment-2616786>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD4VSX7XINVMEWOR676QF5DVGLEYFANCNFSM5RLTCAGA>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>
********************************************************************
This message does not originate from a known Department of Energy email system.
Use caution if this message contains attachments, links or requests for information.

********************************************************************

**Reply by MagnusMyhre** [reply]
Hi Andrew, thanks for your reply.

I did what you said, breaking the loop between the mixer and compressor and this solved at least the initialization heuristic tree and the DOF. 
However when I try to initialize it says in the output log that there is to few degrees of freedom when initializing the heat exchangers. And I receive some evaluation errors including cutting alpha back. Could the reason for this be caused by the fact that my shell outlet is indirectly connected with the tube inlets? Because of this my other unit models get infeasible solutions and I get the initialization error. 

I decided to make 2 property packages, because I had no need for the propane component beside only in two of the unit models. They are identically, but one of them include the Propane component.  Also I replaced the valves with the PressureChanger set with "compressor" = False and "thermodynamically assumption" = Adiabatic

Could you take a look and see if you find a solution? 

Attached is the property packages and the revised model including the last output log(DEBUG). 

[IDAES_model_rev.zip](https://github.com/IDAES/idaes-pse/files/8624775/IDAES_model_rev.zip)

Kind regards,
Magnus

**Comment by andrewlee94** [comment]
Hi Magnus,

Looking at the your code, I think the reason you ae seeing initializing errors with the heat exchangers is because you are fixing the outlet temperatures on both sides of the unit. I see two problems with this:


  1.  First, for HE101 you are defining the temperature at both outlets (shell and tube), but I am fairly certain that is making the problem degenerate. Due to the heat balance on the unit, you can only specify one of the outlet temperatures. The second degree of freedom you need to fix for the heat exchanger is either the area or heat transfer coefficient.
  2.  Secondly, whilst setting the outlet temperature it is a valid way to define the DoF of the heat exchanger, the in-built initialization routines were not built with that in mind (there are so many ways it could be done, unfortunately we couldn’t anticipate all options).

This could well be the reason you are seeing the evaluation errors – a degenerate problem can easily run into evaluation errors.

Andrew

From: MagnusMyhre ***@***.***>
Sent: Wednesday, May 4, 2022 1:39 PM
To: IDAES/idaes-pse ***@***.***>
Cc: Lee, Andrew (FN) ***@***.***>; Comment ***@***.***>
Subject: [EXTERNAL] Re: [IDAES/idaes-pse] Advice on creating model (Discussion #764)


Hi Andrew, thanks for your reply.

I did what you said, breaking the loop between the mixer and compressor and this solved at least the initialization heuristic tree and the DOF.
However when I try to initialize it says in the output log that there is to few degrees of freedom when initializing the heat exchangers. And I receive some evaluation errors including cutting alpha back. Could the reason for this be caused by the fact that my shell outlet is indirectly connected with the tube inlets? Because of this my other unit models get infeasible solutions and I get the initialization error.

I decided to make 2 property packages, because I had no need for the propane component beside only in two of the unit models. They are identically, but one of them include the Propane component. Also I replaced the valves with the PressureChanger set with "compressor" = False and "thermodynamically assumption" = Adiabatic

Could you take a look and see if you find a solution?

Attached is the property packages and the revised model including the last output log(DEBUG).

IDAES_model_rev.zip<https://github.com/IDAES/idaes-pse/files/8624775/IDAES_model_rev.zip>

Kind regards,
Magnus

—
Reply to this email directly, view it on GitHub<https://github.com/IDAES/idaes-pse/discussions/764#discussioncomment-2688918>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD4VSXZTQSIK4IE5GK3MUV3VIKYZZANCNFSM5RLTCAGA>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>
********************************************************************
This message does not originate from a known Department of Energy email system.
Use caution if this message contains attachments, links or requests for information.

********************************************************************

**Reply by MagnusMyhre** [reply]
Ok I see, I will try to unfix the outlet temperatures and add an overall heat transfer coefficient and see if it solves the problem. 

Thanks again.

Magnus

**Comment by MagnusMyhre** [comment]
Hi

I finally got my model to initialize, but there is one problem.
When the initialization of the system is finished, it starts over again until an overflow error shows up like this:
<img width="630" alt="Overflow" src="https://user-images.githubusercontent.com/77294919/167891088-ac684965-7376-44f5-8454-640c26beeff3.png">

Because of this, I have to interrupt the kernel after the last unit model is initialized.

What causes this and how can I fix it? Thanks

Magnus

**Reply by andrewlee94** [reply]
We will need a bit more information on what you are doing to answer this question.

Firstly, as to why the initialization repeats, are using the `SequentialDecomposition` tool and does the new flowsheet have recycle loops? If so, this is expected behavior as it is generally necessary to iteratively solve the recycle loop to get it close enough to solve. You can set a limitation on the number of iterations the tool will do (or alternatively initialize each unit manually in a sequence of your choice).

As for the math range error, we would need to see more details of where the issue is occurring - basically the full traceback you get when it fails (it would also help to know exactly which unit model, property package and initialization step is failing). The short answer is that something is trying to do a calculation that results in an answer outside the range of values supported by Python - e.g. an exponential with a large power.

**Reply by MagnusMyhre** [reply]
As for how the model looks, I decided to cut the arc between the separator and the tube inlets of the heat exchangers. So there are no loops. I do use the SequentialDecomposition tool yes, I was not aware of the iteration procedure here, thanks for the info. 

Attached is the output log(DEBUG) and the model with prop packages for the overflow/math error problem. It occured when initializing the Mixer unit for the second time, here with the property package "HC_PR2".

I also noticed when it lists the sequence of the initialization, it does not show all the unit models, just some of them, not sure why. Still, it initializes all the unit models though. 
[IDAES_model_rev (2).zip](https://github.com/IDAES/idaes-pse/files/8679545/IDAES_model_rev.2.zip)
 
Thanks. 

Kind regards,
Magnus

**Reply by MagnusMyhre** [reply]
I would like to edit my last post: I do have a loop I guess between the heat exchanger tube outlet and shell inlet for "HE102".

Magnus
